dandi/girder.py

```diff
@@ -845,3 +844,99 @@ def __exit__(self, exc_type, exc_value, tb):
             self._pbar.clear()  # remove from screen -- not in effect ATM TODO
             self._pbar.close()
             del self._pbar
+
+
+def keyring_lookup(service_name, username):
+    """
+    Determine a keyring backend to use for storing & retrieving credentials as
+    follows:
+
+    - If the user has specified a backend explicitly via the
+      ``PYTHON_KEYRING_BACKEND`` environment variable or a ``keyringrc.cfg``
+      file, use that backend without checking whether it's usable (If it's not,
+      the user messed up).
+
+    - Otherwise, query the default backend (which is guaranteed to already have
+      the requisite dependencies installed) for the credentials for the given
+      service name and username.  If this completes without error (regardless
+      of whether the backend contains any such credentials), use that backend.
+
+    - If the query fails (e.g., because a GUI is required but the session is in
+      a plain terminal), try using the ``EncryptedKeyring`` backend.
+
+      - If the default backend *was* the ``EncryptedKeyring`` backend, error.
+
+      - If the ``EncryptedKeyring`` backend is not in the list of available
+        backends (likely because its dependencies are not installed, though
+        that shouldn't happen if dandi was installed properly), error.
+
+      - If ``EncryptedKeyring``'s data file already exists, use it as the
+        backend.
+
+      - If ``EncryptedKeyring``'s data file does not already exist, ask the
+        user whether they want to start using ``EncryptedKeyring``.  If yes,
+        then set ``keyringrc.cfg`` (if it does not already exist) to specify it
+        as the default backend, and return the backend.  If no, error.
+
+    Returns a keyring backend and the password it holds (if any) for the given
+    service and username.
+    """
+
+    from keyring.core import get_keyring, load_config, load_env
+    from keyring.backend import get_all_keyring
+    from keyring.errors import InitError
+    from keyring.util.platform_ import config_root
+    from keyrings.alt.file import EncryptedKeyring
```

Comment:
why not to import them all on top?

Reply:
or may be keep them here but we do need some unit (via mockups) and/or integration (do specify `PYTHON_KEYRING_BACKEND` pointing to some pre-crafted file) test to verify this logic.  It would also be sad if we miss the moment when keyring introduces some regression etc and we do not notice since no coverage here